Misleading error on tuple struct field with docstring

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=48a7a316c9d4825be960974fd87a8d8d

```rust
struct Foo {
    a: Vec<(
        /// Docstring
        f32,
        f32,
    )>,
}
```

I expected to see this happen: Add docstring to field in tuple struct

Instead, this happened: Misleading second error "constant provided when a type was expected".

```
   Compiling playground v0.0.1 (/playground)
error: expected type, found doc comment `/// Docstring`
 --> src/lib.rs:3:9
  |
3 |         /// Docstring
  |         ^^^^^^^^^^^^^ expected type
  |
help: expressions must be enclosed in braces to be used as const generic arguments
  |
2 ~     a: Vec<{ (
3 |         /// Docstring
4 |         f32,
5 |         f32,
6 ~     ) }>,
  |

error[E0747]: constant provided when a type was expected
 --> src/lib.rs:2:12
  |
2 |       a: Vec<(
  |  ____________^
3 | |         /// Docstring
4 | |         f32,
5 | |         f32,
6 | |     )>,
  | |_____^
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: aarch64-apple-darwin
release: 1.76.0
LLVM version: 17.0.6
```

--- a/compiler/rustc_parse/src/parser/path.rs
+++ b/compiler/rustc_parse/src/parser/path.rs
@@ -972,7 +972,10 @@ impl<'a> Parser<'a> {
                     GenericArg::Type(ty)
                 }
                 Err(err) => {
-                    if let Some(snapshot) = snapshot
+                    let stopped_at_doc_comment = matches!(self.token.kind, token::DocComment(..));
+
+                    if !stopped_at_doc_comment
+                        && let Some(snapshot) = snapshot
                         && let Some(expr) =
                             self.recover_unbraced_const_arg_that_can_begin_ty(snapshot)
                     {

--- a/tests/ui/parser/doc-comment-in-generic-tuple-type.rs
+++ b/tests/ui/parser/doc-comment-in-generic-tuple-type.rs
@@ -1,3 +1,4 @@
+// Regression test for https://github.com/rust-lang/rust/issues/122463
 struct Foo {
     a: Vec<(
         /// Docstring

--- a/tests/ui/parser/doc-comment-in-generic-tuple-type.rs
+++ b/tests/ui/parser/doc-comment-in-generic-tuple-type.rs
@@ -1,0 +1,10 @@
+struct Foo {
+    a: Vec<(
+        /// Docstring
+        //~^ ERROR expected type, found doc comment
+        f32,
+        f32,
+    )>,
+}
+
+fn main() {}

--- a/tests/ui/parser/doc-comment-in-generic-tuple-type.stderr
+++ b/tests/ui/parser/doc-comment-in-generic-tuple-type.stderr
@@ -1,5 +1,5 @@
 error: expected type, found doc comment `/// Docstring`
-  --> $DIR/doc-comment-in-generic-tuple-type.rs:3:9
+  --> $DIR/doc-comment-in-generic-tuple-type.rs:4:9
    |
 LL | struct Foo {
    |        --- while parsing this struct

--- a/tests/ui/parser/doc-comment-in-generic-tuple-type.stderr
+++ b/tests/ui/parser/doc-comment-in-generic-tuple-type.stderr
@@ -1,0 +1,11 @@
+error: expected type, found doc comment `/// Docstring`
+  --> $DIR/doc-comment-in-generic-tuple-type.rs:3:9
+   |
+LL | struct Foo {
+   |        --- while parsing this struct
+LL |     a: Vec<(
+LL |         /// Docstring
+   |         ^^^^^^^^^^^^^ expected type
+
+error: aborting due to 1 previous error
+

